### ISSUE TITLE
Support code-splitting for UMD builds

### ISSUE THREAD

**Issue by johannes-z** [rollup/rollup#3490]
### Feature Use Case

UMD builds should be able to code-split. Webpack can create code-splitting bundles even in UMD.

#### Input

main.js
```
import('./other')
```
other.js
```
console.log('test')
```

#### Compiled
*main.js*
```
// webpack runtime
__webpack_require__.e(/* import() */ 1).then(__webpack_require__.t.bind(null, "aca4", 7));
```
*other.js*
```
((typeof self !== 'undefined' ? self : this)["webpackJsonpcore"] = (typeof self !== 'undefined' ? self : this)["webpackJsonpcore"] || []).push([[1],{

/***/ "aca4":
/***/ (function(module, exports) {

console.log('test');

/***/ })

}]);
```

### Feature Proposal
rollup should support UMD and code-splitting. The only problem I see here is IIFE bundles. CommonJS and AMD are already supported.

I think webpack can code-split UMD bundles because of their runtime (please correct me if I'm wrong).

In order for rollup to support UMD code-splitting, the split IIFE files have to be loaded and executed somehow (maybe a runtime-helper?). Also I guess all chunks need a UMD factory also.

**Comment by seivan** [comment]
The reason Webpack can do it, is because they have their [own runtime](https://gist.github.com/Rich-Harris/79a02519fb00837e8d5a4b24355881c0#file-webpack-bundle-js) for loading stuff.

Rollup supports `systemjs` but doesn't include its "loader", but you can do that yourself by adding `SystemJS` and necessary polyfills. 

```js
import 'unfetch'
import s from "systemjs/dist/s"
```
Build that as `IIFE` and add its output as a script tag.

**Comment by johannes-z** [comment]
I think the runtime to support this could be rather small. code-splitting UMD would just require another UMD wrapper for the split file. I've written a small plugin for my use-case, that does exactly this.

Sometimes you cannot use SystemJS, as you are locked in by the ecosystem to use something else. Also the "modern browser" argument doesn't hold for production, as in a business environment IE and the old Edge still (sadly) dominate.

If rollup would support code-splitting in these scenarios, you won't really need webpack anymore...

**Comment by seivan** [comment]
> I think the runtime to support this could be rather small. 

You need to load them somehow, which means its going to need polyfills for e.g `Promise` and/or `Fetch`

>  Also the "modern browser" argument doesn't hold for production, as in a business environment IE and the old Edge still (sadly) dominate.

Hence the suggestion of `SystemJS`. 

> Sometimes you cannot use SystemJS, as you are locked in by the ecosystem to use something else.


Right that's a good point, but the way I understand it, and I could be wrong, the only way UMD actually works out of the box is if it defaults to `IIFE`, otherwise you're going to need a runtime loader. You're still better of with `SystemJS`, but then again it's a 2kb polyfill that you could skip if IIFE supported code splits.

I tackled this issue earlier, so I thought I'd offer an alternative.

**Comment by johannes-z** [comment]
Rollup could at least support code-splitting on a named module basis. I wrote a plugin that replaces a dynamic import call with an UMD-factory for named modules:

```js
/**
 * Replaces named, dynamic imports with an UMD factory. Absolute and relative imports are ignored.
 *
 * Requires an environment that supports `Promise`.
 */
export default function pluginDynamicImports (options = {}) {
  return {
    name: 'dynamic-imports',
    transform (code, filename) {
      // TODO: warn/error if globalName is not specified in UMD environment.
      // TODO: global is hardcoded to window, which is wrong.
      const transformedCode = code.replace(/import\(['"`](?![./])(.*?)['"`]\)/gi, (match, request) => {
        const globalName = options.globals[request]
        return `new Promise(function (resolve, reject) {
          (function (global) {
            typeof exports === 'object' && typeof module !== 'undefined' ? resolve(require("${request}")) :
            typeof define === 'function' && define.amd ? require(["${request}"], resolve, reject) :
            (global = global || self, resolve(global["${globalName}"]));
          }(window));
        })`
      })

      return transformedCode
    },
  }
}
```

It assumes that the imported module has an alias set in the `requirejs.config`, and that it exists on the window for an IIFE import.

This works for my use cases, but surely can be improved. It allows me to use code like this without any issues:

```
async function doStuff () {
  if (...) return
  const get = await import('lodash.get')
  get(...)
  // do more stuff
}
```

`doStuff` is for example a very rarely used function, and splitting `lodash.get` would improve initial load by a lot.

**Comment by seivan** [comment]
@johannes-z Out of curiosity, do you know if there are any benefits with `UMD` over `SystemJS`? 

The only thing I can tell is that `SystemJS` is 3x smaller than `RequireJS`  and builtin into Rollup, I can't see any benefit here other the fact that you seem forced to use UMD? I'm actually wondering we should also switch.

**Comment by johannes-z** [comment]
I'm not 100% sure, but in production I only had problems using `SystemJS`. I've worked with `RequireJS` for years and never had an issue, or an issue that I couldn't solve. Interestingly, Microsoft used `SystemJS` for their new SharePoint Modern UI, but after roughly 2 years they changed to `RequireJS`. `RequireJS` and `AMD` seems to be much more flexible. If you are in control of the app and environment you're using/programming, I guess there is not much difference feature wise...

Yes I'm aware of that: I'd actually only use `AMD`, but I don't want to lose the option to just import my scripts using a plain `script`-tag... 

So my point is: Support lazy loading where it is supported (AMD), but not where it doesn't have to be (CommonJS, IIFE). Rollup breaks the build though, which isn't a necessity as this can be circumvented with the plugin I just posted. Rollup could do the same, but warn the user instead of erroring.

**Comment by itssumitrai** [comment]
There is a similar issue https://github.com/rollup/rollup/issues/2072 which was closed a while back. I would also like this to be supported by Rollup out of the box, maybe this could be some config of some sort (like turn on rollup loader if needed).

**Comment by ChristianMurphy** [comment]
A workaround that may work for some babel+rollup users in the meantime, a babel plugin that can translate some dynamic imports to static imports to allow rollup to generate a UMD build.
https://www.npmjs.com/package/babel-plugin-transform-dynamic-imports-to-static-imports

**Comment by andreapigatto** [comment]
@ChristianMurphy do you have any example of this? After including this babel library rollup should be able to split the UMD build automatically? Using Vite (rollup in background) and is able to code splitting only ES files but not UMD files in bundle, I guess cause not supported by Rollup

**Comment by ChristianMurphy** [comment]
@andreapigatto It was also written for an earlier version of rollup (v1) than vite is using (v2).

I do not have an example in vite.

You're welcome to try it out with vite to see if it works, here's an example of how the plugin can be used with plain rollup v1 which may be able to be adapted https://github.com/elmsln/WCFactory/blob/master/packages/rollup-umd-build/rollup.config.factory.mjs

**Comment by RobinTail** [comment]
Is there a complete solution for the splitting and run-time loading, @johannes-z ?

**Comment by johannes-z** [comment]
No, I don't think so. But I have been using the plugin I posted for a while.
